[16.0][OU-ADD] website_slides_survey: Migration to 16.0

carlos-lopez-tecnativa opened the pull request:
TT51975

@Tecnativa @pedrobaeza @MiquelRForgeFlow could you please review this?

carlos-lopez-tecnativa committed:
[OU-ADD] website_slides_survey: Migration to 16.0

MiquelRForgeFlow commented:
/ocabot migration website_slides_survey

MiquelRForgeFlow commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,19 @@
+
+
+def _delete_sql_constraints(env):
+    # Delete constraints to recreate it
```
not sure if this is needed at v16. Do you know @pedrobaeza ?

carlos-lopez-tecnativa replied:
I saw a similar case in the `account` module:
https://github.com/OCA/OpenUpgrade/blob/c158fd44fcf15bc090a14c2e50863789bb4657e5/openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py#L133-L135
Why is it not necessary in version 16? Could you please explain it to me?

MiquelRForgeFlow replied:
Because at some point, I don't remember the version, odoo introduced a fix that automatically makes the ORM update these constrains. But let's merge as it is to assure the constrains are reset.

pedrobaeza replied:
I don't remember either when and if the constraint are correctly recreated.

MiquelRForgeFlow commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+# NOTHING TO DO: The field exists in the website_slides module as storable; just changed it to computed.
+
+website_slides_survey / slide.slide              / slide_category (False)        : NEW selection_keys: ['article', 'certification', 'document', 'infographic', 'quiz', 'video'], mode: modify
+# DONE pre-migration: The field exists in the website_slides module, the migration script is done on website_slides
```
```suggestion
# NOTHING TO DO: The field exists in the website_slides module, the migration script is done on website_slides
```

carlos-lopez-tecnativa replied:
Done, thanks

carlos-lopez-tecnativa commented:
To add context to the latest changes: `slide_type` is set to `null` after the recompute triggered by `rename_columns`
https://github.com/OCA/OpenUpgrade/blob/c158fd44fcf15bc090a14c2e50863789bb4657e5/openupgrade_scripts/scripts/website_slides/16.0.2.6/pre-migration.py#L29

Before the migration process (database from V15):
![image](https://github.com/user-attachments/assets/e0787cb8-6efe-4db1-91e7-d0fb1b934008)

After migrating to V16 without this change, the records are not set to `null` because the computed field is inherited from another module:
https://github.com/odoo/odoo/blob/acbd2da7828090753a9ee3dc85c44bc8eca2aaa8/addons/website_slides_survey/models/slide_slide.py#L69

![image](https://github.com/user-attachments/assets/25278494-af70-4f0f-8932-02e67ce80660)

After migrating to V16 with this change:
![image](https://github.com/user-attachments/assets/2691cf61-7b49-4f89-b21b-d72b8ce18b99)

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,50 @@
+
+from openupgradelib import openupgrade
+
+_translations_to_delete = [
```
Being strict, these are used as constants, so they can be put in upper and with a meaningful name, but you can pass directly the value in the method call instead of putting it here.

```suggestion
XMLIDS_TO_DELETE_TRANSLATIONS = [
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def _delete_sql_constraints(env):
+    # Delete constraints to recreate it
```
It can be a docstring.

```suggestion
    """Delete constraints to recreate it"""
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,50 @@
+        env.cr,
+        """
+        UPDATE slide_slide
+        SET slide_type = slide_category
```
To be more verbose:

```suggestion
        SET slide_type = 'certification'
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+# NOTHING TO DO: The field exists in the website_slides module, the migration script is done on website_slides
+
+website_slides_survey / slide.slide              / slide_type (False)            : selection_keys is now '['article', 'certification', 'doc', 'google_drive_video', 'image', 'pdf', 'quiz', 'sheet', 'slides', 'vimeo_video', 'youtube_video']' ('['certification', 'document', 'infographic', 'presentation', 'quiz', 'video', 'webpage']')
+# Done: Fill the correct value when slide_category=certification
```
```suggestion
# DONE: pre-migration: Filled with "certification" value when slide_category=certification
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+---XML records in module 'website_slides_survey'---
+DEL ir.actions.act_window: website_slides_survey.survey_survey_action_slides_report
+DEL ir.ui.menu: website_slides_survey.website_slides_menu_report_certification
+# NOTHING TO DO: Handle by ORM
```
```suggestion
# NOTHING TO DO: Handled by ORM
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+NEW ir.actions.act_window.view: website_slides_survey.survey_survey_action_slides_view_form
+NEW ir.actions.act_window.view: website_slides_survey.survey_survey_action_slides_view_kanban
+NEW ir.actions.act_window.view: website_slides_survey.survey_survey_action_slides_view_tree
+# NOTHING TO DO: New features
```
```suggestion
# NOTHING TO DO: it's noupdate=0 stuff
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+
+ir.model.constraint: website_slides_survey.constraint_slide_slide_check_certification_preview (changed definition: is now 'check(slide_category != 'certification' or is_preview = false)' ('check(slide_type != 'certification' or is_preview = false)'))
+ir.model.constraint: website_slides_survey.constraint_slide_slide_check_survey_id (changed definition: is now 'check(slide_category != 'certification' or survey_id is not null)' ('check(slide_type != 'certification' or survey_id is not null)'))
+# DONE pre-migration: safely delete constraint to recreate it
```
```suggestion
# DONE: pre-migration: safely delete constraints for letting ORM to recreate them
```

pedrobaeza commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,34 @@
+NEW ir.ui.view: website_slides_survey.survey_survey_view_kanban
+NEW ir.ui.view: website_slides_survey.survey_survey_view_search_slides
+NEW ir.ui.view: website_slides_survey.survey_survey_view_tree_slides
+# NOTHING TO DO: New features
```
This can be grouped with the previous ir noupdate=0 stuff.

pedrobaeza commented:
> To add context to the latest changes: `slide_type` is set to `null` after the recompute triggered by `rename_columns`
> 
> https://github.com/OCA/OpenUpgrade/blob/c158fd44fcf15bc090a14c2e50863789bb4657e5/openupgrade_scripts/scripts/website_slides/16.0.2.6/pre-migration.py#L29
> 
> Before the migration process (database from V15): ![image](https://private-user-images.githubusercontent.com/174346082/440901086-e0787cb8-6efe-4db1-91e7-d0fb1b934008.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NTk3NjAsIm5iZiI6MTc0NjU1OTQ2MCwicGF0aCI6Ii8xNzQzNDYwODIvNDQwOTAxMDg2LWUwNzg3Y2I4LTZlZmUtNGRiMS05MWU3LWQwZmIxYjkzNDAwOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxOTI0MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMDJiMTFiMDUzMGFiMDI3ZWVkNTg0NzVkYmY2NzhkMzViY2Q0MDJmMTQ0ZmM0N2YwYjQ4ZDE5YmQwMzQ4YTllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ciueMBQzJk83j9DGA3Ne2GkVFJmazL3CKmQIFKC0eG8)
> 
> After migrating to V16 without this change, the records are not set to `null` because the computed field is inherited from another module: https://github.com/odoo/odoo/blob/acbd2da7828090753a9ee3dc85c44bc8eca2aaa8/addons/website_slides_survey/models/slide_slide.py#L69
> 
> ![image](https://private-user-images.githubusercontent.com/174346082/440901158-25278494-af70-4f0f-8932-02e67ce80660.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NTk3NjAsIm5iZiI6MTc0NjU1OTQ2MCwicGF0aCI6Ii8xNzQzNDYwODIvNDQwOTAxMTU4LTI1Mjc4NDk0LWFmNzAtNGYwZi04OTMyLTAyZTY3Y2U4MDY2MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxOTI0MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YjM4ZmQyOTQ2NjUwYTRlOWEzNGUyNDRhZDgzMTlhMjQwODA4MmRmMmE1ZDNhMzU3YzNkZTNkOWRhMGZmMzUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.agJzK8U6hmcMM7EuMAoUUUqsQKLKGurK7B_Db8_I1zY)
> 
> After migrating to V16 with this change: ![image](https://private-user-images.githubusercontent.com/174346082/440901207-2691cf61-7b49-4f89-b21b-d72b8ce18b99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NTk3NjAsIm5iZiI6MTc0NjU1OTQ2MCwicGF0aCI6Ii8xNzQzNDYwODIvNDQwOTAxMjA3LTI2OTFjZjYxLTdiNDktNGY4OS1iMjFiLWQ3MmI4Y2UxOGI5OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxOTI0MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iM2M2NzI1OTEyMjhkM2U2YWJkODE5M2NjNmQwYzBiZWNjYzMwMzJkZWI3NjcwNzkzODU3YzgxMDkwMmEyMzQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._U8eyK1F9RfVlRnxMOTubTDqDIpSiJU1xsIEZWtmzRQ)

Maybe it's better to not renaming, and force in end-migration script the overall recomputation?

carlos-lopez-tecnativa committed:
[OU-IMP] website_slides: Finish migration scripts
Recompute slide_type in the end-migration instead of just renaming the column. This prevents incorrect recomputation when only this module is present, and ensures the field is properly recomputed at the end, taking into account other modules such as website_slides_survey.

carlos-lopez-tecnativa commented:
> Maybe it's better to not renaming, and force in end-migration script the overall recomputation?

I’ve updated the code, please let me know if it’s OK for you.

MiquelRForgeFlow commented on openupgrade_scripts/scripts/website_slides_survey/16.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,27 @@
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.load_data(
+        env.cr, "website_slides_survey", "16.0.1.0/noupdate_changes.xml"
+    )
+    openupgrade.delete_record_translations(
+        env.cr,
+        "website_slides_survey",
+        ["mail_template_user_input_certification_failed"],
+    )
```
`load_data` (+`delete_record_translations`) is always called in post-migration.

carlos-lopez-tecnativa replied:
Thanks so much for the correction. I'll take it into account next time.

pedrobaeza commented on openupgrade_scripts/scripts/website_slides/16.0.2.6/end-migration.py:
```diff
@@ -0,0 +1,13 @@
+from openupgradelib import openupgrade
+
+
+def _recompute_slide_slide_type(env):
```
For having a trace of what's going on.

```suggestion
@openupgrade.logging()
def _recompute_slide_slide_type(env):
```

carlos-lopez-tecnativa replied:
Done
